idseq_dag/steps/generate_coverage_viz.py

```diff
@@ -0,0 +1,241 @@
+                    accession_data[values[3]]["reads"].append(values[0])
+
+        for accession_id, data in accession_data.items():
+            accession_data[accession_id]["contigs"] = list(data["contigs"])
+
+        return (accession_data, taxons_to_accessions)
+
+
+    def augment_accession_data_with_info(self, info_dict, accession_data):
+        for accession_id in accession_data:
+            entry = info_dict.get(accession_id)
+
+            if entry:
+                accession_data[accession_id]["name"] = entry[0]
+                accession_data[accession_id]["total_length"] = int(entry[1])
+            else:
+                accession_data[accession_id]["name"] = "Unknown accession"
+                accession_data[accession_id]["total_length"] = 0
+
+
+    def generate_contigs_map(self, blast_top_m8, valid_contigs_with_read_counts):
+        contigs = {}
+
+        # File is empty.
+        if os.path.getsize(blast_top_m8) < MIN_M8_FILE_SIZE:
+            return contigs
+
+        # iterate_m8 automatically removes invalid hits.
+        for _contig_id, _accession_id, _percent_id, _alignment_length, _e_value, _bitscore, line in m8.iterate_m8(blast_top_m8):
+            parts = line.split("\t")
+            name_parts = parts[0].split("_")
+
+            if parts[0] in valid_contigs_with_read_counts:
+                contigs[parts[0]] = {
+                    "total_length": int(name_parts[3]),
+                    "accession": parts[1],
+                    "query_start": int(parts[6]),
+                    "query_end": int(parts[7]),
+                    "subject_start": int(parts[8]),
+                    "subject_end": int(parts[9]),
+                    "prop_mismatch": int(parts[4]) / int(parts[3]),
+                    "percent_id": float(parts[2]),
+                    "alignment_length": int(parts[3]),
+                    "num_mismatches": int(parts[4]),
+                    "num_gaps": int(parts[5]),
+                    "num_reads": valid_contigs_with_read_counts[parts[0]]
+                }
+
+        return contigs
+
+
+    # We process gsnap.deduped.m8 instead of gsnap.reassigned.m8 because we ignore contigs with read_count < 4.
+    # However, these contigs still get reassigned in gsnap.reassigned.m8,
+    # and overwrite the original read alignment to the accession, which we want.
+    def generate_reads_map(self, deduped_m8, unassigned_reads_set):
+        reads = {}
+
+        # File is empty.
+        if os.path.getsize(deduped_m8) < MIN_M8_FILE_SIZE:
+            return contigs
+
+        # iterate_m8 automatically removes invalid hits.
+        for _read_id, _accession_id, _percent_id, _alignment_length, _e_value, _bitscore, line in m8.iterate_m8(deduped_m8):
+            parts = line.split("\t")
+
+            if parts[0] in unassigned_reads_set:
+                reads[parts[0]] = {
+                    # Extract the length from the contig name.
```

Comment:
Why the code duplication? Can make a helper to parse the general m8 into this structure